auth/auth/auth.py

```diff
@@ -263,6 +272,66 @@ async def callback(request):
     return aiohttp.web.HTTPFound(next_page)
 
 
+@routes.post('/api/v1alpha/user/{user}/create')
+@rest_authenticated_developers_only
+async def create_user(request: web.Request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    username = request.match_info['user']
+
+    body = await request.json()
+    login_id = body['login_id']
+    is_developer = body['is_developer']
+    is_service_account = body['is_service_account']
+
+    if not username:
+        raise web.HTTPBadRequest(body='username must be a non-zero length string')
+    if not isinstance(login_id, str) or not login_id:
+        raise web.HTTPBadRequest(body='login_id must be a non-zero length string')
+    if not isinstance(is_developer, bool):
+        raise web.HTTPBadRequest(body='invalid value specified for is_developer')
+    if not isinstance(is_service_account, bool):
+        raise web.HTTPBadRequest(body='invalid value specified for is_service_account')
+    if is_developer and is_service_account:
+        raise web.HTTPBadRequest(body='user cannot be both a developer and a service account')
+
+    existing_users = [
+        x
+        async for x in db.select_and_fetchall(
+            "SELECT * FROM users WHERE login_id = %s OR username = %s;", (login_id, username)
+        )
+    ]
+    if len(existing_users) == 1:
+        existing_user = existing_users[0]
+        expected_username = existing_user['username']
+        expected_login_id = existing_user['login_id']
+        if username != expected_username:
+            raise web.HTTPBadRequest(body=f"user '{expected_username}' already exists with login id '{login_id}'")
+        if login_id != expected_login_id:
+            raise web.HTTPBadRequest(
+                body=f"user '{username}' already exists with a different login id '{expected_login_id}'"
+            )
+    elif len(existing_users) == 0:
+        await db.execute_insertone(
+            '''
+        INSERT INTO users (state, username, login_id, is_developer)
+        VALUES (%s, %s, %s, %s);
+        ''',
+            ('creating', username, login_id, is_developer),
+        )
+    else:
+        raise web.HTTPInternalServerError()
+
+    delay = 5
+    while True:
+        user = await user_from_login_id(db, login_id)
+        if user['state'] != 'creating':
+            break
+        delay = await sleep_and_backoff(delay, max_delay=30)
```

Comment:
We've seen cases where the user doesn't move out of a 'creating' state (and we have to check the Hail batch logs directly) eg, this might hang until a timeout is reached:

- A quota has been reached for project number <project-id>: Service accounts per project
- The service account name was too long.

Reply:
(I think another case was about invalid characters in the service account name (e.g. `michael.franklin` having a dot in the user name, which then got translated to an invalid service account email address), but maybe that has already been fixed.)